stagecoach/price-pusher/price_pusher/orchestrator.py

```diff
@@ -0,0 +1,159 @@
+import logging
+import asyncio
+
+from typing import List, Optional
+
+from pragma.core.entry import Entry, SpotEntry, FutureEntry
+from pragma.core.assets import AssetType, PragmaAsset
+
+from price_pusher.core.poller import PricePoller
+from price_pusher.core.listener import PriceListener
+from price_pusher.core.pusher import PricePusher
+from price_pusher.type_aliases import LatestOrchestratorPairPrices
+from price_pusher.utils.assets import asset_to_pair_id
+
+logger = logging.getLogger(__name__)
+
+
+class Orchestrator:
+    """
+    Main class responsible of the price pushing orchestration.
+    """
+
+    poller: PricePoller
+    listeners: List[PriceListener]
+    pusher: PricePusher
+    # Contains the latest spot/future prices for each sources
+    latest_prices: LatestOrchestratorPairPrices
+    push_queue: asyncio.Queue
+
+    def __init__(
+        self,
+        poller: PricePoller,
+        listeners: List[PriceListener],
+        pusher: PricePusher,
+    ) -> None:
+        # Init class properties.
+        self.poller = poller
+        self.listeners = listeners
+        self.pusher = pusher
+
+        # Contains the latest prices for each sources
+        self.latest_prices: LatestOrchestratorPairPrices = {}
+
+        # Queue containing multiple list of Entries (that will get pushed)
+        self.push_queue = asyncio.Queue()
+
+        # Entities communication.
+        self.poller.set_update_prices_callback(self.callback_update_prices)
+        # Send a reference to the orchestration newest prices to the listeners.
+        # NOTE: the listeners will never mutate the orchestrator prices. Only read.
+        for listener in self.listeners:
+            listener.set_orchestrator_prices(self.latest_prices)
+
+    async def run_forever(self) -> None:
+        """
+        Starts asynchronously the services responsible for the price updates.
+        This include:
+            - the poller, that fetch constantly the latest prices of all pairs
+              from different sources,
+            - the listener, that listen our oracle and push an event when the
+              data is outdated and needs new entries,
+            - the pusher, that pushes entries to our oracle.
+        """
+        await asyncio.gather(
+            self._poller_service(),
+            self._listener_services(),
+            self._pusher_service(),
+        )
+
+    async def _poller_service(self) -> None:
+        """
+        Starts the polling service in its own thread.
+        """
+        while True:
```

Comment:
is that how low we can go?

Reply:
not sure, used an arbitrary number we can probably go lower.

changed to 5 for now and will test more